test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java

```diff
@@ -775,38 +773,62 @@ private List<FieldInjector> injectFields(Class<?> testClass, Object testInstance
     static class FieldInjector {
 
         private final Field field;
-        private final InstanceHandle<?> handle;
+        private final List<InstanceHandle<?>> unsetHandles;
 
         public FieldInjector(Field field, Object testInstance) throws Exception {
-            ArcContainer container = Arc.container();
-            BeanManager beanManager = container.beanManager();
-
             this.field = field;
-            this.handle = container.instance(field.getGenericType(), getQualifiers(field, beanManager));
 
-            if (field.isAnnotationPresent(Inject.class)) {
-                if (handle.getBean().getKind() == io.quarkus.arc.InjectableBean.Kind.SYNTHETIC) {
-                    throw new IllegalStateException(String
-                            .format("The injected field %s expects a real component; but obtained: %s", field,
-                                    handle.getBean()));
+            ArcContainer container = Arc.container();
+            BeanManager beanManager = container.beanManager();
+            java.lang.reflect.Type requiredType = field.getGenericType();
+            Annotation[] qualifiers = getQualifiers(field, beanManager);
+
+            Object injectedInstance;
+
+            if (qualifiers.length == 1 && qualifiers[0].equals(All.Literal.INSTANCE)) {
+                // Special handling for @Injec @All List
```

Comment:
Out of curiosity, what happens if there's another qualifier? `@Inject @All @MyQualifier List<MyType>`

Reply:
That's a very good point and I have to admit that I need to check the code and tests...

Reply:
Ok, so we need to change the logic here to something like `qualifiers.contains(All.Literal.INSTANCE)`... because it should behave as if `Instance<>` was used, i.e. the qualifier on the injection point is taken into account during type-safe resolution. I will modify the test as well.